USQL Migration guide (#1419)
evidence-dev/evidence: sites/docs/docs/guides/usql-migration-guide.md

---
title: Universal SQL Migration Guide
sidebar_position: 3
---

Universal SQL is Evidence's latest release, and fundamentally changes how Evidence queries data in order to bring support for three big new features:
1. **Multiple Data Sources:** Combine data from multiple databases, flat files like CSVs, and even non-SQL data sources like Google Sheets - and query them all with SQL
2. **Inputs and Filters:** Use input components to dynamically update your queries and charts based on user input. Filter data using parameterized queries rather than Javascript filter syntax
3. **Adapter Plugins:** Create your own adapter to enable connecting to any data source

This is a big release and contains several breaking changes. This guide will help you migrate your existing projects to USQL.

> 

Reviewer: Do I need to migrate my project?
```suggestion
## Do I need to migrate my project?
- If your `@evidence-dev/evidence` package version is <=v24.0.0 then it has not yet been migrated, and we encourage you to do so as soon as is practical
- `v23` will continue to be supported for now (sunset date TBC) and will receive critical bugfixes during this period

```

> 

Reviewer: Put a TLDR here about the migration script?

The following steps explain the changes to Evidence with Universal SQL.

If you are trying to migrate an existing project, jump straight to the migration script.

## Breaking Changes
- Existing data source connections will need to be reconfigured
- Queries to your data source will need to be moved to the new `sources` directory (details below)
- Queries in markdown files and in the `queries` directory are now written in DuckDB SQL

> 

Reviewer: Deprecated syntax?
- `{data.my_query}`. Use `{my_query}` instead
- declaring variables from queries in JavaScript `<script>` tags using `let my_var =` or `var my_var =` . Use `$: my_var` instead

## How Universal SQL Works

> - Universal SQL introduces a new data loading step (called `run sources`) - you write queries against multiple data sources, and Evidence will combine them all into Parquet files within a DuckDB WebAssembly instance - so you can query across them using SQL

Reviewer: This is a very technically detailed bullet point. Parquet and WebAssembly are somewhat incidental to the user.

- Once your sources are loaded, they can be queried using DuckDB SQL syntax, either directly on your markdown page, or by writing a `.sql` file in your `queries` directory and referencing it in the frontmatter of your markdown file
- These DuckDB SQL queries run in your browser and can be changed on the fly by interactive user input (e.g., filters, dropdowns) using parameterized queries (e.g., `where customer_id = '${inputs.customer}'`)

## Change to Project Structure
To support multiple data sources, USQL introduces a new `sources` directory, which contains one folder per data source. Each data source folder inclues connection configuration files (defined in YAML, but configurable via the settings UI of your project in your browser).

If you already have a `sources` directory in your project (e.g., for `.sql` files or `.csv` files), it will need to be renamed to `queries`. This functions in the same way as before, but the `.sql` files use DuckDB SQL syntax and the queries run in the browser vs. during the build step.

Queries in your markdown files will use DuckDB SQL syntax and will also run in the browser.

#### Pre-USQL Basic Project Structure
```

> +-- .evidence

Reviewer: ```suggestion
```

necessary?

Author: I think it's okay - just don't want people to think they have a special situation if their folder looks different

+-- pages/
|   `-- index.md
+-- sources/
|   `-- a_query.sql
|   `-- another_query.sql
```

#### USQL Basic Project Structure
```

> +-- .evidence

Reviewer: ```suggestion
```

necessary?

+-- pages/
|   `-- index.md
+-- queries/
|   `-- a_query.sql
|   `-- another_query.sql
+-- sources/
|   `-- my_source/
|        `-- connection.yaml
|        `-- connection.options.yaml
|        `-- source_query.sql
```

Optional folders for `components`, `static`, and `partials` still work as they did before.

## Overview of Changes you will need to make
:::info Recommendations
We recommend creating a backup of your project before starting the migration.
:::

- Use new template project scaffold
- Migrate queries to new project folder structure

> - Update on-page queries to use DuckDB syntax

Reviewer: ```suggestion
- Update on-page queries to use DuckDB syntax
- (Suggested) Refactor `.filter()` statements to take advantage of SQL based filtering
- (If necessary) Update deprecated syntax
- (If neccesary) Reinstall any custom plugins 
- (If necessary) Reinstall any custom external dependencies (non-evidence libraries) 
```

    - Declaring variables from queries in `<script>` tags using `let my_var = ` or `var my_var = `. Use `$: my_var` instead

## How Universal SQL Works
- Universal SQL introduces a new data loading step (called `run sources`) - you write queries against multiple data sources, and Evidence will combine them all into Parquet files within a DuckDB WebAssembly instance - so you can query across them using SQL
- Once your sources are loaded, they can be queried using DuckDB SQL syntax, either directly on your markdown page, or by writing a `.sql` file in your `queries` directory and referencing it in the frontmatter of your markdown file
- These DuckDB SQL queries run in your browser and can be changed on the fly by interactive user input (e.g., filters, dropdowns) using parameterized queries (e.g., `where customer_id = '${inputs.customer}'`)

## Change to Project Structure
To support multiple data sources, USQL introduces a new `sources` directory, which contains one folder per data source. Each data source folder inclues connection configuration files (defined in YAML, but configurable via the settings UI of your project in your browser).

> If you already have a `sources` directory in your project (e.g., for `.sql` files or `.csv` files), it will need to be renamed to `queries`. This functions in the same way as before, but the `.sql` files use DuckDB SQL syntax and the queries run in the browser vs. during the build step.

Reviewer 2: ```suggestion
If you already have a `sources` directory in your project (e.g., for `.sql` files or `.csv` files), it will need to be renamed to `queries`. This functions in the same way as before, but the `.sql` files use DuckDB SQL syntax and the queries run in the browser rather than during the build step.
```

`rather than` might read a bit easier than vs

    - It is normal to see errors in your terminal at this stage - you need to configure your data source to avoid these (see next step)
11. Configure your data source in the Settings menu
    - Click the 3-dot menu at the top right of the page and click Settings
    - In the Data Sources section, click to add a new connection
    - Select the data source type your project uses and provide a name for your data source. This will appear as a folder within the `sources` directory in your project (e.g., you could use `needful_things` if using the Evidence demo database)
    - Click to test your connection
    - Click to confirm the changes
12. Navigate back to the home page in your browser and refresh the page
13. The page should now be working. If you still see errors, they may be related to query chaining or SQL syntax. See the Special Situations section below for information which may help. If you need assistance tracking down the issues, pleace [reach out on Slack](https://slack.evidence.dev)

## Deployment Changes

### Self-Hosting
You will need to update 2 things in your deployment setup to complete the migration to USQL:

> 1. Update your environemt variables

Reviewer 2: ```suggestion
1. Update your environment variables
```

If you use the VS Code migration command, chained queries found on markdown pages are left on the page rather than being moved to the `sources` directory like other queries. This is because we assume that most chained queries are simple enough for the syntax of your source database to match with the DuckDB syntax they will need to move to. In some cases, the syntax will not line up and you will need to make an adjustment.

### Evidence Plugins
If your project includes an Evidence plugin (e.g., [Evidence Labs](https://labs.evidence.dev)):
1. Find the `evidence.plugins.yaml` file in your `_legacy_project` folder and copy the line(s) containing the plugin(s) you're using
2. Paste those lines into the `evidence.plugins.yaml` file in your new project
3. Install the plugin(s) in your project. E.g.,:
```shell
npm install --save @evidence-dev/labs
```

### External Package Dependencies

If your project includes external packages installed via npm, you will need to reinstall those packages so that they are reflected in your `package.json`

```shell
npm install <package-name>

> ```

Reviewer: ````
### Monorepos
If your evidence project is installed in a monorepo then...
````

Author: Added this to the migration steps - ensure you open the evidence project folder specifically. not sure if there's anything else specific to a monorepo we should note?
